README.md
# tap-darksky

> :warning: **The Darksky API Is Shutting Down.** Unfortunately, with the acquisition by Apple, Darksky has [announced](https://blog.darksky.net/) they are no longer taking new signups for the Darksky API.

This is a [Singer](https://singer.io) tap that produces JSON-formatted data
following the [Singer
spec](https://github.com/singer-io/getting-started/blob/master/SPEC.md).

This tap:

- Pulls raw data from the [Darksky Weather API](https://darksky.net/dev/docs#overview)
- Extracts the following resource:
  - [Forecast](https://darksky.net/dev/docs#time-machine-request)
- Outputs the schema for the resource
- Incrementally pulls data based on the input state for each `location`.

## Streams

[forecast](https://darksky.net/dev/docs#time-machine-request)
- Endpoint: https://api.darksky.net/forecast/{SECRET_KEY}/{latitude,longitude},{bookmark-datetime}?exclude=currently,minutely
- Primary key fields: latitude, longitude, forecast_date
- Foreign key fields: None
- Replication strategy: INCREMENTAL (query filtered)
  - Bookmark query fields: forecast_date
  - Bookmark: time (date-time)
- Transformations: de-nest daily data, add start/end times and local date, camelCase to snake_case


## Quick Start

1. Install

    Clone this repository, and then install using setup.py. We recommend using a virtualenv:

    ```bash
    > virtualenv -p python3 venv
    > source venv/bin/activate
    > python setup.py install
    OR
    > cd .../tap-darksky
    > pip install .
    ```
2. Dependent libraries
    The following dependent libraries were installed.
    ```bash
    > pip install singer-python
    > pip install singer-tools
    > pip install target-stitch
    > pip install target-json
    ```
    - [singer-tools](https://github.com/singer-io/singer-tools)
    - [target-stitch](https://github.com/singer-io/target-stitch)

3. Create your tap's `config.json` file. The `start_date` is the absolute minimum date for collecing weather data from your locations. The `user_agent` should list the tap-name and API user email address (for API logging purposes). The `secret_key` may be obtained from [Darksky.net](https://darksky.net/dev) with a free account (limited to 1000 API calls/day) or a subscription account (paying for additional API calls). The `languge` should be the 2-letter code of one of the supported [translation languages](https://github.com/darkskyapp/translations/tree/master/lib/lang); `en` is default.  The `units` should be the 2-letter code for the desired measurement units: `si`, `us`, `auto`, `uk2`, or `ca`; default is `auto`. The `location_list` stores a list of geo-locations in the format with comma separating latitude,longitude and semicolon separating each location. The `location_list` config field in the Stitch UI is limited to text-area config parameters of `varchar(16384)`, or 5461 characters, which is approximately 227 locations (lat,lon; with no spaces and 6-digit decimal precision). The tap will remove any space and non-numerical characters. Format for lists:
`latitude_1,longitude_1;latitude_2,longitude_2;latitude_3,longitude_3;...` 

    ```json
    {
        "secret_key": "YOUR_SECRET_KEY",
        "language": "en",
        "units": "us",
        "location_list": "38.840544, -105.0444233; 45.587467, -122.404503; 45.304104, -121.754761; 39.191097, -106.817535",
        "start_date": "2019-01-01T00:00:00Z",
        "user_agent": "tap-darksky <api_user_email@your_company.com>"
    }
    ```
    **NOTE**: Each `location` for each `forecast_date` is a separate API call. Therefore, `start_date` and  `location_list` values impact the API call usage (and charges, if subscribed). The `free` plan is limited to 1000 API calls/day. The tap will error if the allowed API calls are exceeded for `free` plan accounts.

    Optionally, create a `state.json` file. `currently_syncing` is an optional attribute used for identifying the last object to be synced in case the job is interrupted mid-stream. The next run would begin where the last job left off. For this tap, each `location` is bookmarked. If a new `location` is added to an existing tap, only the new `location` will be synced from the `start_date`. All other `locations` will continue to sync from the `bookmark` date.

    ```json
    {
        "currently_syncing": null,
        "bookmarks": {
            "forecast": {
            "45.304104,-121.754761": "2019-10-04T00:00:00.000000Z",
            "38.840544,-105.0444233": "2019-10-04T00:00:00.000000Z",
            "45.587467,-122.404503": "2019-10-04T00:00:00.000000Z",
            "39.191097,-106.817535": "2019-10-04T00:00:00.000000Z"
            }
        }
    }
    ```

1. Run the Tap in Discovery Mode
    This creates a catalog.json for selecting objects/fields to integrate:
    ```bash
    tap-darksky --config config.json --discover > catalog.json
    ```
   Additionally, the tap provides a query parameter for `excludes` if `daily`, `hourly`, or `flags` are de-selected in Discovery mode.
   See the Singer docs on discovery mode
   [here](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#discovery-mode).

2. Run the Tap in Sync Mode (with catalog) and [write out to state file](https://github.com/singer-io/getting-started/blob/master/docs/RUNNING_AND_DEVELOPING.md#running-a-singer-tap-with-a-singer-target)

    For Sync mode:
    ```bash
    > tap-darksky --config tap_config.json --catalog catalog.json > state.json
    > tail -1 state.json > state.json.tmp && mv state.json.tmp state.json
    ```
    To load to json files to verify outputs:
    ```bash
    > tap-darksky --config tap_config.json --catalog catalog.json | target-json > state.json
    > tail -1 state.json > state.json.tmp && mv state.json.tmp state.json
    ```
    To pseudo-load to [Stitch Import API](https://github.com/singer-io/target-stitch) with dry run:
    ```bash
    > tap-darksky --config tap_config.json --catalog catalog.json | target-stitch --config target_config.json --dry-run > state.json
    > tail -1 state.json > state.json.tmp && mv state.json.tmp state.json
    ```

3. Test the Tap
    
    While developing the darksky tap, the following utilities were run in accordance with Singer.io best practices:
    Pylint to improve [code quality](https://github.com/singer-io/getting-started/blob/master/docs/BEST_PRACTICES.md#code-quality):
    ```bash
    > pylint tap_darksky -d missing-docstring -d logging-format-interpolation -d too-many-locals -d too-many-arguments
    ```
    Pylint test resulted in the following score:
    ```bash
    Your code has been rated at 9.84/10
    ```

    To [check the tap](https://github.com/singer-io/singer-tools#singer-check-tap) and verify working:
    ```bash
    > tap-darksky --config tap_config.json --catalog catalog.json | singer-check-tap > state.json
    > tail -1 state.json > state.json.tmp && mv state.json.tmp state.json
    ```
    Check tap resulted in the following:
    ```bash
    Checking stdin for valid Singer-formatted data
    The output is valid.
    It contained 56 messages for 1 streams.

        1 schema messages
        24 record messages
        31 state messages

    Details by stream:
    +----------+---------+---------+
    | stream   | records | schemas |
    +----------+---------+---------+
    | forecast | 24      | 1       |
    +----------+---------+---------+
    ```
---

Copyright &copy; 2019 Stitch


CHANGELOG.md
# Changelog

## 1.0.1
  * Dependabot update [#6](https://github.com/singer-io/tap-darksky/pull/6)

## 1.0.0
  * Preparing for v1.0.0 release

## 0.0.2
  * Change location_list config parameter from three parameters to one.

## 0.0.1
  * Initial commit

tap_darksky/client.py
import backoff
import requests
from requests.exceptions import ConnectionError
from singer import metrics, utils
from singer.utils import strftime
import singer

LOGGER = singer.get_logger()


class Server5xxError(Exception):
    pass


class Server429Error(Exception):
    pass


class DarkskyError(Exception):
    pass


class DarkskyBadRequestError(DarkskyError):
    pass


class DarkskyUnauthorizedError(DarkskyError):
    pass


class DarkskyPaymentRequiredError(DarkskyError):
    pass


class DarkskyNotFoundError(DarkskyError):
    pass


class DarkskyConflictError(DarkskyError):
    pass


class DarkskyForbiddenError(DarkskyError):
    pass


class DarkskyInternalServiceError(DarkskyError):
    pass


ERROR_CODE_EXCEPTION_MAPPING = {
    400: DarkskyBadRequestError,
    401: DarkskyUnauthorizedError,
    402: DarkskyPaymentRequiredError,
    403: DarkskyForbiddenError,
    404: DarkskyNotFoundError,
    409: DarkskyForbiddenError,
    500: DarkskyInternalServiceError}


def get_exception_for_error_code(status_code):
    return ERROR_CODE_EXCEPTION_MAPPING.get(status_code, DarkskyError)

def raise_for_error(response):
    try:
        response.raise_for_status()
    except (requests.HTTPError, requests.ConnectionError) as error:
        try:
            content_length = len(response.content)
            if content_length == 0:
                # There is nothing we can do here since Darksky has neither sent
                # us a 2xx response nor a response content.
                return
            response_json = response.json()
            status_code = response.status_code
            message = 'RESPONSE: {}'.format(response_json)
            ex = get_exception_for_error_code(status_code)
            raise ex(message)
        except (ValueError, TypeError):
            raise DarkskyError(error)


class DarkskyClient(object):
    def __init__(self,
                 secret_key,
                 user_agent=None):
        self.__secret_key = secret_key
        self.__user_agent = user_agent
        self.__session = requests.Session()
        self.base_url = 'https://api.darksky.net'
        self.__verified = False

    def __enter__(self):
        self.__verified = self.check_secret_key()
        return self

    def __exit__(self, exception_type, exception_value, traceback):
        self.__session.close()

    @backoff.on_exception(backoff.expo,
                          Server5xxError,
                          max_tries=7,
                          factor=3)
    def check_secret_key(self):
        if self.__secret_key is None:
            raise Exception('Error: Missing secret_key.')
        headers = {}
        if self.__user_agent:
            headers['User-Agent'] = self.__user_agent
        headers['Accept'] = 'application/json'
        endpoint = 'forecast'
        location = '38.840544,-105.0444233'
        now_date = strftime(utils.now())[0:10]
        param_string = 'exclude=currently,minutely,hourly,daily,alerts,flags&lang=en&units=us'
        url = '{}/{}/{}/{},{}T00:00:00?{}'.format(
            self.base_url,
            endpoint,
            self.__secret_key,
            location,
            now_date,
            param_string)
        response = self.__session.get(
            # Simple endpoint that returns 1 record (to check API secret_key access):
            url=url,
            headers=headers)
        if response.status_code != 200:
            LOGGER.error('Error status_code = {}'.format(response.status_code))
            raise_for_error(response)
        else:
            resp = response.json()
            if 'latitude' in resp:
                return True
            else:
                return False


    @backoff.on_exception(backoff.expo,
                          (Server5xxError, ConnectionError, Server429Error),
                          max_tries=7,
                          factor=3)
    @utils.ratelimit(800, 60)
    def request(self, method, url, **kwargs):
        if not self.__verified:
            self.__verified = self.check_secret_key()

        url = url.replace('<secret_key>', self.__secret_key)

        if 'endpoint' in kwargs:
            endpoint = kwargs['endpoint']
            del kwargs['endpoint']
        else:
            endpoint = None

        if 'headers' not in kwargs:
            kwargs['headers'] = {}
        kwargs['headers']['Accept'] = 'application/json'

        if self.__user_agent:
            kwargs['headers']['User-Agent'] = self.__user_agent

        if method == 'POST':
            kwargs['headers']['Content-Type'] = 'application/json'

        with metrics.http_request_timer(endpoint) as timer:
            response = self.__session.request(method, url, **kwargs)
            timer.tags[metrics.Tag.http_status_code] = response.status_code

        if response.status_code >= 500:
            raise Server5xxError()

        if response.status_code != 200:
            raise_for_error(response)

        return response.json()

    def get(self, url, **kwargs):
        return self.request('GET', url=url, **kwargs)

    def post(self, url, **kwargs):
        return self.request('POST', url=url, **kwargs)


tap_darksky/schema.py
import os
import json
from singer import metadata
from tap_darksky.streams import STREAMS

# Reference:
# https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#Metadata

def get_abs_path(path):
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)

def get_schemas():
    schemas = {}
    field_metadata = {}

    for stream_name, stream_metadata in STREAMS.items():
        schema_path = get_abs_path('schemas/{}.json'.format(stream_name))
        with open(schema_path) as file:
            schema = json.load(file)
        schemas[stream_name] = schema
        mdata = metadata.new()

        # Documentation:
        # https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#singer-python-helper-functions
        # Reference:
        # https://github.com/singer-io/singer-python/blob/master/singer/metadata.py#L25-L44
        mdata = metadata.get_standard_metadata(
            schema=schema,
            key_properties=stream_metadata.get('key_properties', None),
            valid_replication_keys=stream_metadata.get('replication_keys', None),
            replication_method=stream_metadata.get('replication_method', None)
        )
        field_metadata[stream_name] = mdata

    return schemas, field_metadata


tap_darksky/streams.py
# streams: API URL endpoints to be called
# properties:
#   <root node>: Plural stream name for the endpoint
#   path: API endpoint relative path, when added to the base URL, creates the full path
#   key_properties: Primary key field(s) for the object endpoint
#   replication_method: FULL_TABLE or INCREMENTAL
#   replication_keys: bookmark_field(s), typically a date-time, used for filtering the results
#        and setting the state
#   params: Query, sort, and other endpoint specific parameters
#   data_key: JSON element containing the records for the endpoint
#   bookmark_query_field: Typically a date-time field used for filtering the query
#   bookmark_type: Data type for bookmark, integer or datetime
#   children: A collection of child endpoints (where the endpoint path includes the parent id)
#   parent: On each of the children, the singular stream name for parent element
STREAMS = {
    'forecast':
        {
            'path': 'forecast/<secret_key>/<location>,<forecast_date>',
            'key_properties': ['latitude', 'longitude', 'forecast_date'],
            'replication_method': 'INCREMENTAL',
            'replication_keys': ['forecast_date'],
            'bookmark_type': 'datetime',
            'params': {
                'exclude': '<exclusions_list>',
                'lang': '<language>',
                'units': '<units>'
            }
        }
    }


tap_darksky/__init__.py
#!/usr/bin/env python3

import sys
import json
import argparse
import singer
from singer import metadata, utils
from tap_darksky.client import DarkskyClient
from tap_darksky.discover import discover
from tap_darksky.sync import sync

LOGGER = singer.get_logger()

REQUIRED_CONFIG_KEYS = [
    'secret_key',
    'language',
    'units',
    'location_list',
    'start_date',
    'user_agent'
]

def do_discover():
    LOGGER.info('Starting discover')
    catalog = discover()
    json.dump(catalog.to_dict(), sys.stdout, indent=2)
    LOGGER.info('Finished discover')


@singer.utils.handle_top_exception(LOGGER)
def main():

    parsed_args = singer.utils.parse_args(REQUIRED_CONFIG_KEYS)

    with DarkskyClient(parsed_args.config['secret_key'],
                       parsed_args.config['user_agent']) as client:

        state = {}
        if parsed_args.state:
            state = parsed_args.state

        if parsed_args.discover:
            do_discover()
        elif parsed_args.catalog:
            sync(client=client,
                 config=parsed_args.config,
                 catalog=parsed_args.catalog,
                 state=state)

if __name__ == '__main__':
    main()


tap_darksky/transform.py
import re
import time
import pytz

import singer
from singer.transform import unix_seconds_to_datetime
from singer.utils import strptime_to_utc

LOGGER = singer.get_logger()


# Convert camelCase to snake_case
def convert(name):
    regsub = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', regsub).lower()


# Convert keys in json array
def convert_array(arr):
    new_arr = []
    for i in arr:
        if isinstance(i, list):
            new_arr.append(convert_array(i))
        elif isinstance(i, dict):
            new_arr.append(convert_json(i))
        else:
            new_arr.append(i)
    return new_arr


# Convert keys in json
def convert_json(this_json):
    out = {}
    for key in this_json:
        try:
            new_key = convert(key)
        except TypeError as err:
            LOGGER.error('Error key = {}'.format(key))
            raise err
        if isinstance(this_json[key], dict):
            out[new_key] = convert_json(this_json[key])
        elif isinstance(this_json[key], list):
            out[new_key] = convert_array(this_json[key])
        else:
            out[new_key] = this_json[key]
    return out


def get_min_max_times(this_json):
    time_zone = this_json.get('timezone', 'UTC')
    time_frames = ['daily', 'hourly']
    max_time = 0
    min_time = int(time.time())
    for time_frame in time_frames:
        records = this_json.get(time_frame, {}).get('data', [])
        for record in records:
            rec_time = record.get('time', 0)
            if rec_time > max_time:
                max_time = rec_time
            if rec_time < min_time:
                min_time = rec_time
    # get time in UTC
    utc_dt = strptime_to_utc(unix_seconds_to_datetime(min_time))
    timezone = pytz.timezone(time_zone)
    local_dt = utc_dt.astimezone(timezone)
    local_dt_str = local_dt.strftime('%Y-%m-%d')
    return min_time, max_time, local_dt_str

# de-nest data nodes from daily
def denest_daily_data(this_json):
    new_json = this_json
    # Get first records of single record daily data list
    daily_data = next(iter(this_json.get('daily', {}).get('data', [])), None)
    if daily_data:
        new_json['daily'] = daily_data
    return new_json

# Run all transforms: convert camelCase to snake_case
def transform_json(this_json):
    converted_json = convert_json(this_json)
    start_time, end_time, local_date = get_min_max_times(converted_json)
    converted_json['start_time'] = start_time
    converted_json['end_time'] = end_time
    converted_json['local_date'] = local_date
    converted_json['forecast_date'] = '{}T00:00:00Z'.format(local_date)
    denested_json = denest_daily_data(converted_json)
    return denested_json


tap_darksky/sync.py
from datetime import timedelta
import re
import singer
from singer import metrics, metadata, Transformer, utils, UNIX_SECONDS_INTEGER_DATETIME_PARSING
from singer.utils import strptime_to_utc
from tap_darksky.transform import transform_json
from tap_darksky.streams import STREAMS

LOGGER = singer.get_logger()


def write_schema(catalog, stream_name):
    stream = catalog.get_stream(stream_name)
    schema = stream.schema.to_dict()
    try:
        singer.write_schema(stream_name, schema, stream.key_properties)
    except OSError as err:
        LOGGER.info('OS Error writing schema for: {}'.format(stream_name))
        raise err


def write_record(stream_name, record, time_extracted):
    try:
        singer.messages.write_record(stream_name, record, time_extracted=time_extracted)
    except OSError as err:
        LOGGER.info('OS Error writing record for: {}'.format(stream_name))
        LOGGER.info('record: {}'.format(record))
        raise err
    except TypeError as err:
        LOGGER.info('Type Error writing record for: {}'.format(stream_name))
        LOGGER.info('record: {}'.format(record))
        raise err


def get_bookmark(state, stream, location, default):
    if (state is None) or ('bookmarks' not in state):
        return default
    return (
        state
        .get('bookmarks', {})
        .get(stream, {})
        .get(location, default)
    )


def write_bookmark(state, stream, location, value):
    if 'bookmarks' not in state:
        state['bookmarks'] = {}
    if stream not in state['bookmarks']:
        state['bookmarks'][stream] = {}
    state['bookmarks'][stream][location] = value
    LOGGER.info('Write state for Stream: {}, Location: {}, value: {}'.format(
        stream, location, value))
    singer.write_state(state)


def transform_datetime(this_dttm):
    with Transformer() as transformer:
        new_dttm = transformer._transform_datetime(this_dttm)
    return new_dttm


def process_records(catalog, #pylint: disable=too-many-branches
                    stream_name,
                    records,
                    time_extracted,
                    bookmark_field=None,
                    max_bookmark_value=None,
                    last_datetime=None):
    stream = catalog.get_stream(stream_name)
    schema = stream.schema.to_dict()
    stream_metadata = metadata.to_map(stream.metadata)

    with metrics.record_counter(stream_name) as counter:
        for record in records:
            # Transform record for Singer.io
            with Transformer(integer_datetime_fmt=UNIX_SECONDS_INTEGER_DATETIME_PARSING) \
                as transformer:
                transformed_record = transformer.transform(
                    record,
                    schema,
                    stream_metadata)

                # Reset max_bookmark_value to new value if higher
                if transformed_record.get(bookmark_field):
                    if max_bookmark_value is None or \
                        transformed_record[bookmark_field] > transform_datetime(max_bookmark_value):
                        max_bookmark_value = transformed_record[bookmark_field]

                if bookmark_field and (bookmark_field in transformed_record):
                    last_dttm = transform_datetime(last_datetime)
                    bookmark_dttm = transform_datetime(transformed_record[bookmark_field])
                    # Keep only records whose bookmark is after the last_datetime
                    if bookmark_dttm:
                        if bookmark_dttm >= last_dttm:
                            write_record(stream_name, transformed_record, \
                                time_extracted=time_extracted)
                            counter.increment()
                else:
                    write_record(stream_name, transformed_record, time_extracted=time_extracted)
                    counter.increment()

        return max_bookmark_value, counter.value


# Sync a specific parent or child endpoint.
def sync_endpoint(client,
                  catalog,
                  state,
                  start_date,
                  stream_name,
                  url,
                  location,
                  bookmark_field=None):

    # Get the latest bookmark for the stream and set the last_integer/datetime
    last_datetime = get_bookmark(state, stream_name, location, start_date)
    max_bookmark_value = last_datetime

    end_dttm = utils.now()
    end_dt = end_dttm.date()
    start_dttm = strptime_to_utc(last_datetime)
    start_dt = start_dttm.date()

    # date_list provides one date for each date in range
    date_list = [str(start_dt + timedelta(days=x)) for x in range((end_dt - start_dt).days + 1)]

    total_records = 0
    for bookmark_date in date_list:
        LOGGER.info('Stream: {}, Syncing bookmark_date = {}'.format(
            stream_name, bookmark_date))
        forecast_date = '{}T00:00:00'.format(bookmark_date)
        forecast_url = url.replace('<forecast_date>', forecast_date)
        LOGGER.info('URL for Stream {}: {}'.format(stream_name, forecast_url))

        # API request data
        data = {}
        data = client.get(
            url=forecast_url,
            endpoint=stream_name)

        # time_extracted: datetime when the data was extracted from the API
        time_extracted = utils.now()
        if not data or data is None or data == {}:
            break # No data results

        # Transform data with transform_json from transform.py
        # LOGGER.info('data = {}'.format(data)) # TESTING, comment out
        transformed_data = [] # initialize the record list
        transformed_data.append(transform_json(data))

        # LOGGER.info('transformed_data = {}'.format(transformed_data)) # TESTING, comment out
        if not transformed_data or transformed_data is None or transformed_data == []:
            LOGGER.info('No transformed data for data = {}'.format(data))
            break # No data results

        # Process records and get the max_bookmark_value and record_count for the set of records
        max_bookmark_value, record_count = process_records(
            catalog=catalog,
            stream_name=stream_name,
            records=transformed_data,
            time_extracted=time_extracted,
            bookmark_field=bookmark_field,
            max_bookmark_value=max_bookmark_value,
            last_datetime=last_datetime)

        total_records = total_records + record_count
        LOGGER.info('Stream {}, location: {}, batch processed {} records'.format(
            stream_name, location, record_count))

        # Update the state with the max_bookmark_value for the stream
        if bookmark_field:
            write_bookmark(state, stream_name, location, max_bookmark_value)

    # Return total_records (for all pages)
    return total_records


# Currently syncing sets the stream currently being delivered in the state.
# If the integration is interrupted, this state property is used to identify
#  the starting point to continue from.
# Reference: https://github.com/singer-io/singer-python/blob/master/singer/bookmarks.py#L41-L46
def update_currently_syncing(state, stream_name):
    if (stream_name is None) and ('currently_syncing' in state):
        del state['currently_syncing']
    else:
        singer.set_currently_syncing(state, stream_name)
    singer.write_state(state)


def sync(client, config, catalog, state):
    start_date = config.get('start_date')
    language = config.get('language', 'en')
    units = config.get('units', 'auto')

    # Stitch UI config parameters may be text areas, but limited to varchar(16384), 5461 chars,
    #   approx. 227 lat,lon locations (without spaces, 6-digit precision)
    # locations_list should be delimited like: lat1,lon1;lat2,lon2;lat3,lon3
    #   commas separating lat,lon and semicolons ; separating locations
    locations = config.get('location_list')

    # Remove non-numeric chars except commas, semicolons, minus signs
    #   and Split to list of locations by semicolon
    location_list = re.sub('[^0-9,.;-]', '', locations).split(";")

    # Get selected_streams from catalog, based on state last_stream
    #   last_stream = Previous currently synced stream, if the load was interrupted
    last_stream = singer.get_currently_syncing(state)
    LOGGER.info('last/currently syncing stream: {}'.format(last_stream))
    selected_streams = []
    for stream in catalog.get_selected_streams(state):
        selected_streams.append(stream.stream)
    LOGGER.info('selected_streams: {}'.format(selected_streams))

    if not selected_streams or selected_streams == []:
        return

    # Loop through endpoints in selected_streams
    for stream_name, endpoint_config in STREAMS.items():
        if stream_name in selected_streams:
            LOGGER.info('START Syncing: {}'.format(stream_name))
            update_currently_syncing(state, stream_name)
            write_schema(catalog, stream_name)
            bookmark_field = next(iter(endpoint_config.get('replication_keys', [])), None)
            # Get exclusions based on catalog not selected
            exclusions_list = 'currently,minutely'
            if stream_name == 'forecast':
                # Create exclusions_list from catalog breadcrumb
                stream = catalog.get_stream(stream_name)
                mdata = metadata.to_map(stream.metadata)
                sections_all = ['hourly', 'daily', 'flags']
                for section in sections_all:
                    if not singer.metadata.get(mdata, ('properties', section), 'selected'):
                        # metadata is selected for the dimension
                        exclusions_list = '{},{}'.format(exclusions_list, section)

            # Replace in url and query params
            params = endpoint_config.get('params', {})
            # Squash params and replace language and units
            querystring = '&'.join(['%s=%s' % (key, value) for (key, value) in \
                params.items()]).replace('<language>', language).replace(
                    '<units>', units).replace('<exclusions_list>', exclusions_list)
            endpoint_total = 0
            # Loop for each lat,lon location in location_list
            for location in location_list:
                loc_path = endpoint_config.get('path').replace('<location>', location)
                url = '{}/{}?{}'.format(
                    client.base_url,
                    loc_path,
                    querystring)
                total_records = sync_endpoint(
                    client=client,
                    catalog=catalog,
                    state=state,
                    start_date=start_date,
                    stream_name=stream_name,
                    url=url,
                    location=location,
                    bookmark_field=bookmark_field)

                update_currently_syncing(state, None)
                LOGGER.info('FINISHED Syncing: {}, location: {}, total_records: {}'.format(
                    stream_name,
                    location,
                    total_records))
                endpoint_total = endpoint_total + total_records

            LOGGER.info('FINISHED Syncing: {}, total_records: {}'.format(
                stream_name,
                endpoint_total))
